src/scheduling/flow/cpu_cost_model.cc

```diff
@@ -315,7 +329,32 @@ FlowGraphNode* CpuCostModel::GatherStats(FlowGraphNode* accumulator,
   CHECK_NOTNULL(other->rd_ptr_);
   ResourceDescriptor* rd_ptr = accumulator->rd_ptr_;
   CHECK_NOTNULL(rd_ptr);
-  if (accumulator->type_ == FlowNodeType::MACHINE) {
+  if (accumulator->type_ == FlowNodeType::PU) {
+    CHECK(other->resource_id_.is_nil());
+    ResourceStats latest_stats;
+    bool have_sample = knowledge_base_->GetLatestStatsForMachine(other->resource_id_,
+                                                                 &latest_stats);
+    if (have_sample) {
+      VLOG(2) << "Updating PU " << accumulator->resource_id_ << "'s "
+              << "resource stats!";
+      // Get the CPU stats for this PU
+      string label = rd_ptr->friendly_name();
+      uint64_t idx = label.find("PU #");
+      if (idx != string::npos) {
+        string core_id_substr = label.substr(idx + 4, label.size() - idx - 4);
+        uint32_t core_id = strtoul(core_id_substr.c_str(), 0, 10);
+        float available_cpu_cores =
+                latest_stats.cpus_stats(core_id).cpu_capacity() *
+                (1.0 - latest_stats.cpus_stats(core_id).cpu_utilization());
+        rd_ptr->mutable_available_resources()->set_cpu_cores(
+                available_cpu_cores);
+      }
+      // Running/idle task count
+      rd_ptr->set_num_running_tasks_below(rd_ptr->current_running_tasks_size());
+      rd_ptr->set_num_slots_below(FLAGS_max_tasks_per_pu);
```

Comment:
Should not we set the rd.num_slots_below()?

Reply:
Yes. We need to have alternative way of setting this number to number of allocatable  pods on that particular core(PU). If not core, at least we need to get to know this number for every scheduling round. So that we can remove this constant number. And because of 'FLAGS_max_tasks_per_pu' currently we can't schedule more than 'FLAGS_max_tasks_per_pu' number of pods on that PU or corresponding Machine.

So possible alternative would be,
1) Use very big number which is always greater than allocatable pods for any node.
or
2) Get number of allocatable  pods from poseidon or any stats server like heapster, and use it to replace FLAGS_max_tasks_per_pu.

So in this way, we don't restrict flow from Machine to PU by this number FLAGS_max_tasks_per_pu. So that we can schedule more pods than FLAGS_max_tasks_per_pu for each PU.

I am verifying this part by putting debugs in firmament code, will share more details soon.

Reply:
Current firmament code fixed the maximum number of pods that can be scheduled on PU by using constant number FLAGS_max_tasks_per_pu. So because of this when running_tasks on PU equals FLAGS_max_tasks_per_pu, capacity of arc from Machine to PU is being set to zero. So we are restricted to schedule only FLAGS_max_tasks_per_pu number of tasks/pods on the PU. So in order to remove this restriction we need to set this rd.num_slots_below() to maximum pods that can be scheduled on that PU, that too only for PU.

Code changes can be like.
1)Add new field ‘max_pods’ in ResourceDescriptor, which gets value from kubelet parameter ‘max-pods’ only once when machine is added.
2)For PU node only, while updating the resource descriptor, update num_slots_below to max_pods like below.
rd.set_num_slots_below(machine_rd.max_pods());

In this way, we can schedule max_pods number of pods on that PU not just FLAGS_max_tasks_per_pu.

Reply:
Currently `max-pods` is passed to `num_slots_below`.  Need a new field?

Reply:
Since num_slots_below gets changed every scheduling round, This value does not persist. So better to add new field.